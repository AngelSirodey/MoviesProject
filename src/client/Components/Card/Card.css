.Card__container {
    background-color: #ffffff;
    display: flex;
    flex-direction: column;
    box-shadow: 2px 2px 4px #888888;
    border-radius: 6px;
    width: 280px;
    height: 460px;
    margin: 12px;
}

.Card__image {
    width: 100%;
    height: 75%;
    border-radius: 6px 6px 0px 0px;
}

.Card__description {
    display: flex;
    flex-direction: column;
    height: 100%;
    color: #000000;
}

.Card__title {
    font-size: 22px;
    margin: 5px 10px 0px 10px;
    text-align: start;
}

.Card__links {
    display: flex;
    font-size: 14px;
    font-style: italic;
    justify-content: space-around;
    margin-top: auto;
    padding-bottom: 4px;
}

.Card__links > p {
    margin: 0;
    cursor: pointer;
}

.Card__links > a {
    color: #000000;
    display: block;
    padding: 0;
    text-decoration: none;
}


@media screen and (max-width: 449px) {
    .Card__container {
        width: 220px;
        height: 360px;
    }

    .Card__title {
        font-size: 16px;
    }

    .Card__links {
        font-size: 12px;
    }
}
