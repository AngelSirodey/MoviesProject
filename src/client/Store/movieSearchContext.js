import React, { useState, useContext } from 'react';

import { getMovieByName } from '../../server';

export const MoviesContext = React.createContext();
export const FavoritesContext = React.createContext();

const MoviesProvider = ({children}) => {
	const storageFaves = JSON.parse(localStorage.getItem('user')) || [];
	const [searchResult, setSearchResult] = useState([]);
	const [userFavorites, setUserFavorites] = useState(storageFaves);
	
	return <MoviesContext.Provider value={{ search: [searchResult, setSearchResult], favorites: [userFavorites, setUserFavorites]}}>{children}</MoviesContext.Provider>;
};

const useMovieContext = () => {
	const { search, favorites } = useContext(MoviesContext);
	const [searchResult, setSearchResult] = search;
	const [userFavorites, setUserFavorites] = favorites;

	const saveFavorites = (movieCard) => {
		const userF = userFavorites;
		userF.push(movieCard);
		setUserFavorites(userF);
		localStorage.setItem('user', JSON.stringify(userF));
	};

	const deleteFavorites = (movieCard) => {
		const userF = userFavorites;
		const indexToDelete = userF.findIndex(movie => movie.imdbID === movieCard.imdbID);
		userF.splice(indexToDelete, 1);
		localStorage.setItem('user', JSON.stringify(userFavorites));
		setUserFavorites(userF);
	};

	const searchMovie = debounce((title) => {
        getMovieByName(title).then(setSearchResult)
    }, 500);

    return {
		searchResult: searchResult.Search,
		searchError: searchResult.Error,
		searchTotalResults: searchResult.totalResults,
		searchMovie,
		saveFavorites,
		userFavorites,
		deleteFavorites
	};
};

function debounce(func, wait, immediate) {
	let timeout;
	return function() {
		const context = this, args = arguments;
		const later = function() {
			timeout = null;
			if (!immediate) func.apply(context, args);
		};
		const callNow = immediate && !timeout;
		clearTimeout(timeout);
		timeout = setTimeout(later, wait);
		if (callNow) func.apply(context, args);
	};
};

export { MoviesProvider, useMovieContext };
