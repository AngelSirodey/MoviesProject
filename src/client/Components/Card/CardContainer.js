import React from 'react';

import Card from './Card';

const CardContainer = ({ movies, fromFavorite }) => {
    debugger
    return movies ? (
        movies.map( movie => <Card key={movie.imdbID} movie={movie} fromFavorite={fromFavorite} />)
    ) : null
};

export default CardContainer;
