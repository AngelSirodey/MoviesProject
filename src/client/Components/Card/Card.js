import React from 'react';
import PropTypes from 'prop-types';
import { Link } from "react-router-dom";

import { useMovieContext } from '../../Store/movieSearchContext';

import './Card.css';

const Card = ({ movie, fromFavorite, handleDeleteFavorites }) => {
  const { saveFavorites } = useMovieContext();
  const { Poster, Title, imdbID } = movie;
  const detailsURL = `/details?imdbID=${imdbID}`;

   return (
      <div className="Card__container">
        <img src={Poster} className="Card__image" alt="poster" />
        <div className="Card__description">
          <p className="Card__title">{Title}</p>
          <div className="Card__links">
            {!fromFavorite ?
              <p onClick={() => saveFavorites(movie)} >Add to favotires</p> :
              <p onClick={() => handleDeleteFavorites(movie)} >Remove</p>}
           <Link to={detailsURL}> More details -></Link>
          </div>
        </div>
      </div>
   );
};

Card.propTypes = {
  fromFavorite: PropTypes.bool
};

Card.defaultProps = {
  fromFavorite: false
};

export default Card;
