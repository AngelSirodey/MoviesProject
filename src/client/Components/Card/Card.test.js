import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Card from './Card';
import { MoviesProvider } from '../../Store/movieSearchContext';

Enzyme.configure({ adapter: new Adapter() });


let wrapper;

const movie = {
    Poster: 'image1.jpg',
    Title: 'title1',
    imdbID: '1'
  };

describe('Card Component behaviour', () => {
    it('should render as expected', () => {
        wrapper = shallow(
        <MoviesProvider>
            <Card movie={movie} search="search" />
        </MoviesProvider>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
