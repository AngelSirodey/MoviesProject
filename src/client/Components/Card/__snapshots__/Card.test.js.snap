// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component behaviour should render as expected 1`] = `ShallowWrapper {}`;
